src/lib.rs

```diff
@@ -0,0 +1,155 @@
+    /// It is recommended to panic if it is possible to know that deadlock has occurred.
+    ///
+    /// For implementations on a system with threads, blocking may be the correct choice.
+    ///
+    /// For implementations where a single thread is present, panicking immediately may be
+    /// the correct choice.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
+}
+
+/// Blocking mutex (not async)
+///
+/// Provides a blocking mutual exclusion primitive backed by an implementation of [`ScopedRawMutex`].
+///
+/// Which implementation you select depends on the context in which you're using the mutex, and you can choose which kind
+/// of interior mutability fits your use case.
+///
+/// Use [`CriticalSectionRawMutex`] when data can be shared between threads and interrupts.
+///
+/// Use [`LocalRawMutex`] when data is only shared between tasks running on the same executor.
+///
+/// Use [`ThreadModeRawMutex`] when data is shared between tasks running on the same executor but you want a global singleton.
+///
+/// In all cases, the blocking mutex is intended to be short lived and not held across await points.
+///
+/// [`CriticalSectionRawMutex`]: crate::impls::cs::CriticalSectionRawMutex
+/// [`LocalRawMutex`]: crate::impls::local::LocalRawMutex
+/// [`ThreadModeRawMutex`]: crate::impls::thread_mode::ThreadModeRawMutex
+pub struct BlockingMutex<R, T: ?Sized> {
+    // NOTE: `raw` must be FIRST, so when using ThreadModeMutex the "can't drop in non-thread-mode" gets
+    // to run BEFORE dropping `data`.
+    raw: R,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<R: ScopedRawMutex + Send, T: ?Sized + Send> Send for BlockingMutex<R, T> {}
+unsafe impl<R: ScopedRawMutex + Sync, T: ?Sized + Send> Sync for BlockingMutex<R, T> {}
+
+impl<R: ConstInit, T> BlockingMutex<R, T> {
+    /// Creates a new mutex in an unlocked state ready for use.
+    #[inline]
+    pub const fn new(val: T) -> BlockingMutex<R, T> {
+        BlockingMutex {
+            raw: R::INIT,
+            data: UnsafeCell::new(val),
+        }
+    }
+}
+
+impl<R: ScopedRawMutex, T> BlockingMutex<R, T> {
+    /// Locks the raw mutex and grants temporary access to the inner data
+    ///
+    /// Behavior when the lock is already locked is dependent on the behavior
+    /// of the Raw mutex. See [`ScopedRawMutex::lock()`]'s documentation for
+    /// more details
+    pub fn lock<U>(&self, f: impl FnOnce(&mut T) -> U) -> U {
+        self.raw.lock(|| {
+            let ptr = self.data.get();
+            // SAFETY: Raw Mutex proves we have exclusive access to the inner data
+            let inner = unsafe { &mut *ptr };
+            f(inner)
+        })
+    }
+
+    /// Locks the raw mutex and grants temporary access to the inner data
+    ///
+    /// Returns `Some(U)` if the lock was obtained. Returns `None` if the lock
+    /// was already locked
+    #[must_use]
+    pub fn try_lock<U>(&self, f: impl FnOnce(&mut T) -> U) -> Option<U> {
```

Comment:
perhaps these should have `#[track_caller]` so that if the raw-mutex impl panics due to a deadlock, it's clearer where in the user code the deadlock occurred? this is especially valuable on no-std targets without backtraces.

Reply:
This would be a non-breaking change, so it's fine to address it later.